libc: Default to picolibc where supported

keith-packard opened the pull request:
This switches the default C library to picolibc for all targets which support it.

Closes https://github.com/zephyrproject-rtos/zephyr/issues/49922

keith-packard committed:
libc: Default to picolibc where supported

This switches the default C library to picolibc for all targets which
support it.

Signed-off-by: Keith Packard <keithp@keithp.com>

carlescufi commented:
@keith-packard labeling this as TSC since this requires a bit more discussion.

evgeniy-paltsev commented on lib/libc/Kconfig:
```diff
@@ -55,8 +55,8 @@ menu "C Library"
 	prompt "C Library Implementation"
 	default EXTERNAL_LIBC if NATIVE_APPLICATION
+	default PICOLIBC
 	default NEWLIB_LIBC if REQUIRES_FULL_LIBC
-	default PICOLIBC if REQUIRES_FULL_LIBC
 	default MINIMAL_LIBC
```
I'd say that we need to have better undestanding of our strategy about Zephyr minimal libc and newlib support.

* minimal libc:
As per #49922 picolibc is now supported on all Zephyr architectures (correct me if I'm wrong), so when this change will be applied we'll stop test anyting with minimal libc in upstream CI.

  As I understand we still have plans to keep Zephyr minimal libc - so we need it to have it tested in CI.

* newlib:
It's not clear if we are going to support (and do testing) both picolibc and newlib (for SDK's toolchain) simultaneously or we are going to switch for picolibc for both use cases (if minimal libc implementation is required and if full-sized libc required).

  As some tests select newlib directly in `prj.conf` by setting `CONFIG_NEWLIB_LIBC=y` and not `CONFIG_REQUIRES_FULL_LIBC=y` currently we'll still do some testing in upstream CI with newlib even after this PR is applied. Though currently it's not clear if we are going to keep it that way or we are going to modify these tests to select picolibc by default (as we already mark picolibc as a full-sized libc).

keith-packard replied:
> The idea is that the kernel tests will still build with the minimal libc to ensure that the Zephyr kernel can still build using the minimal libc even if we switch to Picolibc as the default -- this is to be implemented (tracked by "Devise a process to ensure that the Zephyr kernel can keep building with the minimal libc" in #49922).

Do we want to do this by creating explicit tests that select minimal libc (as is currently done for newlib and picolibc)? Or do we want a separate mechanism to enable another CI pipeline that runs tests against minimal libc instead of picolibc?

> That needs to be changed to use `CONFIG_REQUIRES_FULL_LIBC=y` unless the test/sample **must** use newlib.

We also have tests for libc functionality that currently run against all of the SDK C libraries. Expanding that to cover more of the C library functionality used by Zephyr would also serve to help get better C library testing for all of them.

keith-packard replied:
It sounds like we need to change some kernel tests to explicitly test against the minimal C library? I'd think those would want to test against both that and the default C library? That seems easy enough to manage.

keith-packard replied:
I've expanded the kernel tests to run against both the default C library as well as the minimal C library.

keith-zephyr commented on lib/libc/Kconfig:
```diff
@@ -55,8 +55,8 @@ menu "C Library"
+	default PICOLIBC
 	default NEWLIB_LIBC if REQUIRES_FULL_LIBC
-	default PICOLIBC if REQUIRES_FULL_LIBC
 	default MINIMAL_LIBC
```
Should the `default MINIMAL_LIBC` be removed with change, or have some other criteria applied?

stephanosio replied:
This is still relevant because `MINIMAL_LIBC` should be used if `PICOLIBC` is not supported by the toolchain in use.

evgeniy-paltsev replied:
Currently it's used with ARC MWDT toolchain (in case of no full-size libc is required). So we still need ti have it in the end of the list.

stephanosio commented:
As per https://github.com/zephyrproject-rtos/zephyr/issues/49922#issuecomment-1533407667, this needs more documentation and process-related work before it can be approved by the TSC.

I will look into that and create a new PR based on this with the necessary changes.

cfriedt reviewed (commented):
Looks reasonable to me. @keith-packard - are the linker script changes more or less independent of picolibc? If so, we could probably move those out to a separate PR to simplify this one.

keith-packard commented:
> Looks reasonable to me. @keith-packard - are the linker script changes more or less independent of picolibc? If so, we could probably move those out to a separate PR to simplify this one.

sorry, I've been testing #57727 by merging it into this PR. I'm expecting that to get merged and then this one will get rebased on top. The only actual change that *should* be in this PR is the change to lib/libc/Kconfig ...

stephanosio commented:
The general direction of making Picolibc the default libc for Zephyr, as proposed in #49922, has been approved by the TSC on 2023-05-17.

stephanosio commented:
> that there is significant increase in compiled files with `CONFIG_PICOLIBC=y` is in prj.conf file. May be my Zephyr SDK is old?

Have you tried with `CONFIG_PICOLIBC_MODULE=n`, which forces the toolchain-bundled Picolibc to be used?

lawrencek52 commented:
I am using Zephyr SDK 0.16.0. My zephyr is close to the tip of 3.3.99. My target board is xiao_ble_sense.  I compiled "hello_world" with and without  picolibc and there is no significant difference in the binary size. however the number of files compiled jumps from 180 to 1090.

I then compiled a slightly more complex project in the same board.
With CONFIG_PICOLIBC=y:
[1095/1095] Linking C executable zephyr/zephyr.elf
Memory region         Used Size  Region Size  %age Used
           FLASH:       50596 B       788 KB      6.27%
             RAM:       21516 B       256 KB      8.21%
        IDT_LIST:          0 GB         2 KB      0.00%

and Without CONFIG_PICOLIBC:
[185/185] Linking C executable zephyr/zephyr.elf
Memory region         Used Size  Region Size  %age Used
           FLASH:       47236 B       788 KB      5.85%
             RAM:         21 KB       256 KB      8.20%
        IDT_LIST:          0 GB         2 KB      0.00%

as you can see code size in this case is 3k larger. With more a more complex project I saw even larger size differences. I am not sure what is going on here....

stephanosio commented:
> however the number of files compiled jumps from 180 to 1090.

@lawrencek52 Please see https://github.com/zephyrproject-rtos/zephyr/pull/57340#issuecomment-1551642511.

> I then compiled a slightly more complex project in the same board.

It depends on which function(s) from Picolibc you are using (note that unused functions get "garbage collected" and removed from the final binary).

We will need to identify which functions are causing the code size increase. You can follow the steps below to do a footprint comparison:

1. Compile your project using the `footprint` target (e.g. `make footprint` or `ninja footprint`); once using the minimal libc and once using the Picolibc.
2. Use `scripts/footprint/fpdiff.py` to compare the footprint (post the output of this script). For example, `scripts/footprint/fpdiff.py build_minlibc/rom.json build_picolibc/rom.json`.

keith-packard commented:
> as you can see code size in this case is 3k larger. With more a more complex project I saw even larger size differences. I am not sure what is going on here....

That's very surprising; I've never seen it build a *larger* binary. Feel free to attach .map and .config files from the two builds and I can take a look. Maybe there's some dependency that is pulling in the picolibc floating point printf code and not pulling in the minimal C library floating point printf code?

lawrencek52 commented:
There is nothing 'magic' in my source code, here is is so you can replicate:
[Xiao_strip_leds.tgz](https://github.com/zephyrproject-rtos/zephyr/files/11500664/Xiao_strip_leds.tgz)

keith-packard commented:
> There is nothing 'magic' in my source code, here is is so you can replicate: [Xiao_strip_leds.tgz](https://github.com/zephyrproject-rtos/zephyr/files/11500664/Xiao_strip_leds.tgz)

ok, I've figured out the cause of the increase -- the picolibc floating point printf code is quite a bit larger than the minimal C library floating point printf code, so if you enable `CBPRINTF_FP_SUPPORT`, then you'll see a larger resulting binary as cbprintf is implemented in terms of picolibc's fprintf code. Disable that option and the binary size using picolibc is slightly smaller than with the minimal C library.

We can file this as a bug against picolibc if you like, but I don't think we can keep picolibc meeting the ISO/IEC C library specifications while saving this much text space. That specification requires that floating point conversions be 'round tripable' where printing and then scanning a value will always generate a value matching the original one. Picolibc does have a mode where it can perform imprecise conversions, but that code hasn't seen as much optimization work and so it's not actually much smaller. 

We could adopt the minimal C library imprecise floating point output implementation inside picolibc; that should save considerable space when building with imprecise output support (which is not the default).

lawrencek52 commented:
I think as long as it is documented this is fine. The entry for CONFIG_CBPRINTF_FP_SUPPORT should point out the difference between using MINIMAL_LIB and PICOLIBC, in terms of code size, and accuracy of  the numbers printed.

I believe that the changeover to PICOLIBC going to affect the code size of any program using CBPRINTF_FP_SUPPORT.

keith-packard commented:
I found an actual bug in picolibc's cmake support -- it doesn't respect the exact floating point conversion option from Zephyr, `CONFIG_PICOLIBC_IO_FLOAT_EXACT`. Further, the inexact code was using floating point operations which would pull in the soft float library. I've got two PRs pending for picolibc, https://github.com/picolibc/picolibc/pull/470 and https://github.com/picolibc/picolibc/pull/471. Once those are merged, if you disable exact float io and disable c99 printf support, the picolibc build is about 300 bytes smaller than the minimal C library build.

keith-zephyr commented on tests/arch/x86/pagetables/testcase.yaml:
```diff
@@ -3,3 +3,5 @@ tests:
     tags: userspace mmu
     filter: CONFIG_MMU
+    extra_configs:
+      - CONFIG_COMMON_LIBC_MALLOC_ARENA_SIZE=0
```
This may not be needed if you can set the malloc arena size to 0 for PICOLIBC.

keith-packard replied:
see above.

keith-packard commented on samples/userspace/syscall_perf/prj.conf:
```diff
@@ -4,3 +4,4 @@ CONFIG_ASSERT=y
 CONFIG_LOG=y
 CONFIG_LOG_MODE_MINIMAL=y
 CONFIG_MAIN_STACK_SIZE=2048
+CONFIG_COMMON_LIBC_MALLOC_ARENA_SIZE=0
```
Yeah, the problem is that existing picolibc-based applications would all get to change their configuration to specify the arena size.

keith-zephyr replied:
Ack - I see that there are 70+ uses of PICOLIBC in the samples/tests.

keith-packard commented:
> p.s. We will make `config PICOLIBC_USE_MODULE` `default n` to make the toolchain-bundled Picolibc the default option when available.

Note that this probably needs to depend on the various Kconfig build options for picolibc being set to the default values -- if a user overrides those, then presumably they'll want to use the module.

keith-packard commented:
> Note that this probably needs to depend on the various Kconfig build options for picolibc being set to the default values -- if a user overrides those, then presumably they'll want to use the module.

Well, that certainly can't work -- all of those options are hidden if `PICOLIBC_USE_MODULE` isn't selected. Ok, apps wanting a non-default picolibc build will need to select *both* `PICOLIBC_USE_MODULE` and whatever compilation modifiers they need.

keith-packard commented:
I've trimmed this PR down to just the picolibc configuration switch without including any of the required fixes. Those are all on separate PRs and are linked from #49922

nashif commented:
@keith-packard  playing with this PR and xtensa clang toolchain, getting this:

```
/home/nashif/zephyrproject/modules/lib/picolibc/newlib/libc/string/memmove.c:71:1: error: malformed 'optimize' attribute '-fno-tree-loop-distribute-patterns' [-Werror,-Wattribute-optimize]
__inhibit_loop_to_libcall
^
/home/nashif/zephyrproject/zephyr/twister-out/intel_adsp_ace15_mtpm/samples/philosophers/sample.kernel.philosopher.coop_only/modules/picolibc/picolibc/include/sys/cdefs.h:563:33: note: expanded from macro '__inhibit_loop_to_libcall'
  __attribute__ ((__optimize__ ("-fno-tree-loop-distribute-patterns")))
                                ^
1 error generated.
```

keith-packard commented:
> 



> @keith-packard playing with this PR and xtensa clang toolchain, getting this:
> 
> ```
> /home/nashif/zephyrproject/modules/lib/picolibc/newlib/libc/string/memmove.c:71:1: error: malformed 'optimize' attribute '-fno-tree-loop-distribute-patterns' [-Werror,-Wattribute-optimize]
> __inhibit_loop_to_libcall
> ^
> /home/nashif/zephyrproject/zephyr/twister-out/intel_adsp_ace15_mtpm/samples/philosophers/sample.kernel.philosopher.coop_only/modules/picolibc/picolibc/include/sys/cdefs.h:563:33: note: expanded from macro '__inhibit_loop_to_libcall'
>   __attribute__ ((__optimize__ ("-fno-tree-loop-distribute-patterns")))
>                                 ^
> 1 error generated.
> ```

That's odd -- the cmake script should be auto-detecting compiler support for this flag and not using it unless it works. Check out the configuration of picolibc and see what it does when compiling cmake/have-cc-inhibit-loop-to-libcall.c during the cmake process.

nashif commented:
I see this:

```
  -
    kind: "try_compile-v1"
    backtrace:
      - "/home/nashif/zephyrproject/modules/lib/picolibc/cmake/picolibc.cmake:150 (try_compile)"
      - "/home/nashif/zephyrproject/modules/lib/picolibc/CMakeLists.txt:192 (picolibc_flag)"
    directories:
      source: "/home/nashif/zephyrproject/zephyr/build/modules/picolibc/CMakeFiles/CMakeTmp"
      binary: "/home/nashif/zephyrproject/zephyr/build/modules/picolibc/CMakeFiles/CMakeTmp"
    cmakeVariables:
      CMAKE_C_FLAGS: ""
      CMAKE_C_FLAGS_DEBUG: "-g"
      CMAKE_EXE_LINKER_FLAGS: ""
      CMAKE_MODULE_PATH: "/home/nashif/zephyrproject/zephyr/cmake/modules"
    buildResult:
      variable: "_HAVE_CC_INHIBIT_LOOP_TO_LIBCALL"
      cached: true
      stdout: |
        Change Dir: /home/nashif/zephyrproject/zephyr/build/modules/picolibc/CMakeFiles/CMakeTmp

        Run Build Command(s):/usr/bin/ninja -v cmTC_38e9c && [1/2] /home/nashif/xtensa/XtDevTools/install/tools/RI-2022.10-linux/XtensaTools/bin/xt-
clang   -Werror=missing-declarations -Werror=unknown-attributes -Werror=attributes -Werror=unsupported-floating-point-opt -Werror=ignored-optimizati
on-argument -Wno-unused-command-line-argument -MD -MT CMakeFiles/cmTC_38e9c.dir/have-cc-inhibit-loop-to-libcall.c.obj -MF CMakeFiles/cmTC_38e9c.dir/
have-cc-inhibit-loop-to-libcall.c.obj.d -o CMakeFiles/cmTC_38e9c.dir/have-cc-inhibit-loop-to-libcall.c.obj -c /home/nashif/zephyrproject/modules/lib
/picolibc/cmake/have-cc-inhibit-loop-to-libcall.c
        warning: unknown warning option '-Werror=unsupported-floating-point-opt'; did you mean '-Werror=unsupported-target-opt'? [-Wunknown-warning-
option]
        /home/nashif/zephyrproject/modules/lib/picolibc/cmake/have-cc-inhibit-loop-to-libcall.c:1:31: warning: malformed 'optimize' attribute '-fno-
tree-loop-distribute-patterns' [-Wattribute-optimize]
        __attribute__ ((__optimize__ ("-fno-tree-loop-distribute-patterns")))
                                      ^
        2 warnings generated.
        [2/2] : && /home/nashif/.local/lib/python3.10/site-packages/cmake/data/bin/cmake -E rm -f libcmTC_38e9c.a && /home/nashif/xtensa/XtDevTools/
install/tools/RI-2022.10-linux/XtensaTools/bin/xt-ar qc libcmTC_38e9c.a  CMakeFiles/cmTC_38e9c.dir/have-cc-inhibit-loop-to-libcall.c.obj && /home/na
shif/xtensa/XtDevTools/install/tools/RI-2022.10-linux/XtensaTools/bin/xt-ranlib libcmTC_38e9c.a && :

      exitCode: 0
```

keith-packard commented:
> ```
> clang   -Werror=missing-declarations -Werror=unknown-attributes -Werror=attributes -Werror=unsupported-floating-point-opt -Werror=ignored-optimizati
> on-argument -Wno-unused-command-line-argument -MD -MT CMakeFiles/cmTC_38e9c.dir/have-cc-inhibit-loop-to-libcall.c.obj -MF CMakeFiles/cmTC_38e9c.dir/
> have-cc-inhibit-loop-to-libcall.c.obj.d -o CMakeFiles/cmTC_38e9c.dir/have-cc-inhibit-loop-to-libcall.c.obj -c /home/nashif/zephyrproject/modules/lib
> /picolibc/cmake/have-cc-inhibit-loop-to-libcall.c
>         warning: unknown warning option '-Werror=unsupported-floating-point-opt'; did you mean '-Werror=unsupported-target-opt'? [-Wunknown-warning-
> option]
>         /home/nashif/zephyrproject/modules/lib/picolibc/cmake/have-cc-inhibit-loop-to-libcall.c:1:31: warning: malformed 'optimize' attribute '-fno-
> tree-loop-distribute-patterns' [-Wattribute-optimize]
>         __attribute__ ((__optimize__ ("-fno-tree-loop-distribute-patterns")))
> ```
> 
> Is the check ignoring the warning and considering this as a supported attribute?

Looks like that version of clang uses a different warning label for unknown optimization attributes? Maybe try hacking in `-Werror=attribute-optimize` and see if that helps? I can't run this toolchain as it requires a license(!) and there's no way I could sign that.

keith-packard commented:
> > ```
> > clang   -Werror=missing-declarations -Werror=unknown-attributes -Werror=attributes -Werror=unsupported-floating-point-opt -Werror=ignored-optimizati
> > on-argument -Wno-unused-command-line-argument -MD -MT CMakeFiles/cmTC_38e9c.dir/have-cc-inhibit-loop-to-libcall.c.obj -MF CMakeFiles/cmTC_38e9c.dir/
> > have-cc-inhibit-loop-to-libcall.c.obj.d -o CMakeFiles/cmTC_38e9c.dir/have-cc-inhibit-loop-to-libcall.c.obj -c /home/nashif/zephyrproject/modules/lib
> > /picolibc/cmake/have-cc-inhibit-loop-to-libcall.c
> >         warning: unknown warning option '-Werror=unsupported-floating-point-opt'; did you mean '-Werror=unsupported-target-opt'? [-Wunknown-warning-
> > option]
> >         /home/nashif/zephyrproject/modules/lib/picolibc/cmake/have-cc-inhibit-loop-to-libcall.c:1:31: warning: malformed 'optimize' attribute '-fno-
> > tree-loop-distribute-patterns' [-Wattribute-optimize]
> >         __attribute__ ((__optimize__ ("-fno-tree-loop-distribute-patterns")))
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > Is the check ignoring the warning and considering this as a supported attribute?
> 
> Looks like that version of clang uses a different warning label for unknown optimization attributes? Maybe try hacking in `-Werror=attribute-optimize` and see if that helps? I can't run this toolchain as it requires a license(!) and there's no way I could sign that.

Or maybe just use `-Werror` to cover all the bases?

nashif commented:
tried this, seems to work...

```
--- a/cmake/picolibc.cmake
+++ b/cmake/picolibc.cmake
@@ -124,7 +124,7 @@ _picolibc_supported_compile_options(PICOLIBC_FLAG_OPTIONS
   -Werror=missing-declarations
   -Werror=unknown-attributes
   -Werror=attributes
-  -Werror=unsupported-floating-point-opt
+  -Werror=attribute-optimize
   -Werror=ignored-optimization-argument
   -Wno-unused-command-line-argument
   )
```
not sure if this was what you wanted.
now just getting a few tests not building correctly. will take a close look and report back.

keith-packard commented:
> tried this, seems to work...
> 
> ```
> --- a/cmake/picolibc.cmake
> +++ b/cmake/picolibc.cmake
> @@ -124,7 +124,7 @@ _picolibc_supported_compile_options(PICOLIBC_FLAG_OPTIONS
>    -Werror=missing-declarations
>    -Werror=unknown-attributes
>    -Werror=attributes
> -  -Werror=unsupported-floating-point-opt
> +  -Werror=attribute-optimize
>    -Werror=ignored-optimization-argument
>    -Wno-unused-command-line-argument
>    )
> ```
> 
> not sure if this was what you wanted. now just getting a few tests not building correctly. will take a close look and report back.

yup, that's good enough for this test. We'll actually need to leave the other flag here as well; these are all vetted by CMake before being used for compiler feature tests. I've got a patch cooked up, just didn't get time to finish pushing it out; thanks for sorting out the problem on your own in the meantime. Let me know if you figure out the test build failures.

keith-packard committed:
tests/kernel: Also run tests using minimallibc

With picolibc being the default C library, we need to explicitly include
testing against the minimal C library for kernel components.

Signed-off-by: Keith Packard <keithp@keithp.com>

tejlmand reviewed (dismissed):
lgtm :tada:

keith-packard commented:
sorry for the update -- I rebased on top of an unrelated fix which was blocking CI. Now that it's passing, it would be a good time to get this merged :-)

nashif commented on lib/libc/Kconfig:
```diff
@@ -54,8 +54,8 @@ menu "C Library"
+	default PICOLIBC
 	default NEWLIB_LIBC if REQUIRES_FULL_LIBC
-	default PICOLIBC if REQUIRES_FULL_LIBC
 	default MINIMAL_LIBC
```
dont we end up with both picolibc and minimal libc set as default? or does order here play a role?

keith-packard replied:
Kconfig appears to pick the first which matches and which can be enabled -- the goal is to use picolibc where supported otherwise use minimal C library.

keith-packard commented:
> @keith-packard, now that we have a [migration guide](https://github.com/zephyrproject-rtos/zephyr/blob/main/doc/releases/migration-guide-3.5.rst), would you mind adding a note there? I assume under "Required changes", given that users may see their applications break depending on what they are using?

#62261 has some possible things to check. I'm not sure how these should be integrated into the rest of the migration notes though.
